File: freqazoid/src/math/TwoWayMismatch.java
package math;

public class TwoWayMismatch {
	
	private static double pmP = 0.5;
	private static double pmQ = 1.4;
	private static double pmR = 0.5;
	private static double mpP = 0.5;
	private static double mpQ = 1.4;
	private static double mpR = 0.5;
	private static double rho = 0.33;
	
	private static double[] errors1;
	private static double[] ftrials1;
	
	private static double[] errors2;
	private static double[] ftrials2;
	
	public static double calculateFundamentalFrequency(double f0Min, double f0Max, Peak[] peaks) {
		
		// divide the frequency range to equal tempered semitone steps.
		double log102 = 1/Math.log10(2);		
		int N = (int)Math.floor(12*log102*Math.log10(f0Max/f0Min));
		
		errors1 = new double[N];
		ftrials1 = new double[N];
		
		for (int i = 0; i < N; i++) {
			ftrials1[i] = f0Min*Math.pow(2, (double)i/12);
			errors1[i] = calculateTotalError(ftrials1[i], peaks);
		}
		
		// find the one with minimum error
		double minError = Double.MAX_VALUE;		
		int index = 0;
		for (int i = 0; i < N; i++) {
			if(errors1[i] < minError) {
				minError = errors1[i];
				index = i;
			}
		}
		
		
		/*
		 * second time
		 */
		{
			f0Min = ftrials1[index]*Math.pow(2.0, -1.0/12.0);
			f0Max = ftrials1[index]*Math.pow(2.0, 1.0/12.0);
		
			log102 = 1/Math.log10(2);
//			N = (int)Math.floor(10*log102*Math.log10(f0Max/f0Min));
			N = 20;
		
			errors2 = new double[N];
			ftrials2 = new double[N];
		
			for (int i = 0; i < N; i++) {
//				ftrials[i] = f0Min*Math.pow(2, (double)i/10);
				ftrials2[i] = i*(f0Max - f0Min)/N + f0Min;
				errors2[i] = calculateTotalError(ftrials2[i], peaks);
//				System.out.println(ftrials[i]+ ": " + errors[i] );
			}
		
			minError = Double.MAX_VALUE;		
			index = 0;
			for (int i = 0; i < N; i++) {
				if(errors2[i] < minError) {
					minError = errors2[i];
					index = i;
				}
			}
		}
		
//		System.out.println(index + ", " + ftrials[index]);
		return ftrials2[index];
	}
	
	public static double[] getErrors1() {
		return errors1;
	}
	
	public static double[] getFTrials1() {
		return ftrials1;
	}

	private static double calculateTotalError(double f0Trial, Peak[] peaks) {
		double predictedToMeasuredError = 0.0;
		double measuredToPredictedError = 0.0;
		double totalError = 0.0;
		
		double[] measuredPartials = new double[peaks.length];
		double[] measuredAmplitudes = new double[peaks.length];
		
		// determine maximum measured amplitude and partial frequency
		double ampMax = 0.0;
		double freqMax = 0.0;
		for (int i = 0; i < peaks.length; i++) {
			measuredPartials[i] = peaks[i].frequency;
			measuredAmplitudes[i] = peaks[i].amplitude;
			
			if( measuredAmplitudes[i] > ampMax ) {
				ampMax = measuredPartials[i];
			}
			
			if( measuredPartials[i] > freqMax ) {
				freqMax = measuredPartials[i];
			}
		}		
		
		// calculate predicted frequencies of harmonics
		int nPredictedHarmonics = (int)Math.ceil(freqMax/f0Trial);
		double[] predictedHarmonics = new double[nPredictedHarmonics];
		double[] predictedAmplitudes = new double[nPredictedHarmonics];
		for (int n = 0; n < predictedHarmonics.length; n++) {
			predictedHarmonics[n] = f0Trial*(n+1);	
		}
		
		// predicted to measured
		double[] differences = new double[nPredictedHarmonics];
		for (int n = 0; n < differences.length; n++) {
			differences[n] = Double.MAX_VALUE;
			for (int k = 0; k < measuredPartials.length; k++) {
				double diff = Math.abs( predictedHarmonics[n]-measuredPartials[k]);
				if( diff < differences[n] ) {
					differences[n] = diff;
					predictedAmplitudes[n] = measuredAmplitudes[k];
				}
			}
			
//			System.out.print("pe: ");
//			System.out.print( predictedToMeasuredError(differences[n], predictedHarmonics[n], predictedAmplitudes[n], ampMax)+ "  ");
			predictedToMeasuredError += predictedToMeasuredError(differences[n], predictedHarmonics[n], predictedAmplitudes[n], ampMax);
		}
//		System.out.println( predictedToMeasuredError );
		
		// measured to predicted
		differences = new double[peaks.length];
		for (int k = 0; k < differences.length; k++) {
			differences[k] = Double.MAX_VALUE;
			for (int n = 0; n < predictedHarmonics.length; n++) {
				double diff = Math.abs( predictedHarmonics[n]-peaks[k].frequency );
				if( diff < differences[k] ) {
					differences[k] = diff;					
				}
			}
			measuredToPredictedError += measuredToPredictedError(differences[k], peaks[k].frequency, peaks[k].amplitude, ampMax);
		}
		
		totalError = predictedToMeasuredError/predictedHarmonics.length + rho*measuredToPredictedError/peaks.length;
		
//		System.out.println(predictedToMeasuredError/predictedHarmonics.length + ", " + rho*measuredToPredictedError/peaks.length);
//		System.out.println(totalError);
		return totalError;
	}
	
	private static double predictedToMeasuredError(double deltaFreqN, double freqN, double ampN, double ampMax ) {		
		double error = deltaFreqN*Math.pow(freqN, -pmP)+(ampN/ampMax)*(pmQ*deltaFreqN*Math.pow(freqN, -pmP)-pmR);
		// freqN 0 oldugunda error Infinity oluyor.
//		System.out.println(deltaFreqN + ", " + freqN + ", " + ampN + ", " + ampMax);
//		System.out.println(error);
		return error;
	}
	
	private static double measuredToPredictedError(double deltaFreqK, double freqK, double ampK, double ampMax) {
		double error = deltaFreqK*Math.pow(freqK, -mpP)+(ampK/ampMax)*(mpQ*deltaFreqK*Math.pow(freqK, -mpP)-mpR);
		return error;
	}

	public static double getMpP() {
		return mpP;
	}

	public static void setMpP(double mpP) {
		TwoWayMismatch.mpP = mpP;
	}

	public static double getMpQ() {
		return mpQ;
	}

	public static void setMpQ(double mpQ) {
		TwoWayMismatch.mpQ = mpQ;
	}

	public static double getMpR() {
		return mpR;
	}

	public static void setMpR(double mpR) {
		TwoWayMismatch.mpR = mpR;
	}

	public static double getPmP() {
		return pmP;
	}

	public static void setPmP(double pmP) {
		TwoWayMismatch.pmP = pmP;
	}

	public static double getPmQ() {
		return pmQ;
	}

	public static void setPmQ(double pmQ) {
		TwoWayMismatch.pmQ = pmQ;
	}

	public static double getPmR() {
		return pmR;
	}

	public static void setPmR(double pmR) {
		TwoWayMismatch.pmR = pmR;
	}

	public static double getRho() {
		return rho;
	}

	public static void setRho(double rho) {
		TwoWayMismatch.rho = rho;
	}	
}

File: src/twoWayMismatch.js
class PeakDetector {
    constructor() {
        this.peaks = [];
        this.initial = 0.05;
        this.final = 0.004;
        this.freqDecay = 1000.0;
    }

    detectPeaks(spectrum) {
        this.peaks = [];
        var mag = spectrum;
        var dFreq = 22050 / mag.length;
        for(var ix = 1; ix < mag.length - 1; ix++) {
            if(mag[ix] > mag[ix - 1] && mag[ix] > mag[ix + 1]) {
                var freq = ix * dFreq;
                var threshold = this.threshold(freq);
                if(mag[ix] > threshold) {
                    var [dix, vertexMag] = this.parabolaVertex(mag[ix - 1], mag[ix], mag[ix + 1]);
                    this.peaks.push([freq + dix * dFreq, vertexMag]);
                }
            }
        }
    }

    /**
     * Compute x-coordinate of the peak of a parabola (vertex) that passes through these three points:
     * (-1, y1), (-1, y2), (-1, y2)
     *
     * Parabola equation: y = a * x ** 2 + b * x + c
     *
     * x == -1 => a - b + c = y1
     * x ==  0 => c = y2
     * x ==  1 => a + b + c = y3
     *
     * a - b = y1 - y2
     * a + b = y3 - y2
     *
     * which gives:
     *
     * a = 0.5 * (y1 + y3) - y2
     * b = 0.5 * (y3 - y1)
     * c = y2
     *
     * therefore the coordinates of the vertex is: [x, y] = [- b / (2 * a), - b ** 2 / (4 * a) + c]
     *
     * @param {float} y1
     * @param {float} y2
     * @param {float} y3
     */
    parabolaVertex(y1, y2, y3) {
        var a = 0.5 * (y1 + y3) - y2;
        var b = 0.5 * (y3 - y1);
        var x = - 0.5 * b / a;
        var y = - 0.25 * b * b / a + y2;
        return [x, y];
    }

    threshold(freq) {
        return (this.initial - this.final) * Math.exp(- freq / this.freqDecay) + this.final;
    }
}

class TwoWayMismatch {
    constructor() {
        this.fundamentalFrequency = 0.0;
        this.totalError = 0.0;
        this.peakDetector = new PeakDetector();
        this.pmP = 0.5;
        this.pmQ = 1.4;
        this.pmR = 0.5;
        this.mpP = 0.5;
        this.mpQ = 1.4;
        this.mpR = 0.5;
        this.rho = 0.33;

        this.errors1 = [];
        this.ftrials1 = [];
        this.errors2 = [];
        this.ftrials2 = [];

        this.errorThreshold = 7.0;

        this.f0Min = 220;
        this.f0Max = 880;
    }

    process(samples, spectrum) {
        this.peakDetector.detectPeaks(spectrum);
        this.computeFundamentalFrequency();
    }

    computeFundamentalFrequency() {
        var peaks = this.peakDetector.peaks;
        if(peaks.length === 0) {
            this.fundamentalFrequency = null;
            return;
        }

		// divide the frequency range to equal tempered semitone steps.
		var log102 = 1.0 / Math.log10(2.0);
		var nFreqs = Math.floor(12.0 * log102 * Math.log10(this.f0Max / this.f0Min));

		this.errors1 = [];
		this.ftrials1 = [];

		for (var i = 0; i < nFreqs; i++) {
			this.ftrials1[i] = this.f0Min * Math.pow(2.0, i / 12.0);
			this.errors1[i] = this.calculateTotalError(this.ftrials1[i], peaks);
		}

		var minErrorVal = Number.POSITIVE_INFINITY;
		var minErrorIdx = 0;
		for (let i = 0; i < nFreqs; i++) {
			if(this.errors1[i] < minErrorVal) {
				minErrorVal = this.errors1[i];
				minErrorIdx = i;
			}
		}

		// Second pass
        var f0Min2 = this.ftrials1[minErrorIdx] * Math.pow(2.0, -1.0 / 12.0);
        var f0Max2 = this.ftrials1[minErrorIdx] * Math.pow(2.0, 1.0 / 12.0);

        log102 = 1.0 / Math.log10(2.0);
        // nFreqs = Math.floor(12.0 * log102 * Math.log10(f0Max2 /  f0Min2));
        nFreqs = 20;

		this.errors2 = Array(nFreqs);
		this.ftrials2 = Array(nFreqs);

        for (let i = 0; i < nFreqs; i++) {
            // ftrials[i] = f0Min*Math.pow(2, (double)i/10);
            this.ftrials2[i] = i * (f0Max2 - f0Min2) / nFreqs + f0Min2;
            this.errors2[i] = this.calculateTotalError(this.ftrials2[i], this.peaks);
        }

		var minErrorVal = Number.POSITIVE_INFINITY;
		var minErrorIdx = 0;
		for (let i = 0; i < nFreqs; i++) {
			if(this.errors2[i] < minErrorVal) {
				minErrorVal = this.errors2[i];
				minErrorIdx = i;
			}
		}

        this.fundamentalFrequency = this.ftrials2[minErrorIdx];
        this.totalError = minErrorVal;
    }

    calculateTotalError(f0Trial) {
        var peaks = this.peakDetector.peaks;

		var predictedToMeasuredError = 0.0;
		var measuredToPredictedError = 0.0;
		var totalError;

		var measuredPartials = []; // of length peaks
		var measuredAmplitudes = [];

		// determine maximum measured amplitude and partial frequency
		var ampMax = 0.0;
		var freqMax = 0.0;
		for (var i = 0; i < peaks.length; i++) {
            var freq = peaks[i][0];
            var mag = peaks[i][1];
			measuredPartials[i] = freq;
			measuredAmplitudes[i] = mag;

			if( measuredAmplitudes[i] > ampMax ) {
				ampMax = measuredPartials[i];
			}

			if( measuredPartials[i] > freqMax ) {
				freqMax = measuredPartials[i];
			}
		}

		// calculate predicted frequencies of harmonics
		var nPredictedHarmonics = Math.ceil(freqMax / f0Trial);
		var predictedHarmonics = new Float32Array(nPredictedHarmonics);
		var predictedAmplitudes = new Float32Array(nPredictedHarmonics);
		for (let n = 0; n < nPredictedHarmonics; n++) {
			predictedHarmonics[n] = f0Trial * (n + 1);
		}

		// predicted to measured
		var differences = new Float32Array(nPredictedHarmonics);
		for (let n = 0; n < predictedHarmonics.length; n++) {
			differences[n] = Number.POSITIVE_INFINITY;
			for (let k = 0; k < measuredPartials.length; k++) {
				let diff = Math.abs(predictedHarmonics[n] - measuredPartials[k]);
				if(diff < differences[n]) {
					differences[n] = diff;
					predictedAmplitudes[n] = measuredAmplitudes[k];
				}
			}
			predictedToMeasuredError += this.computePredictedToMeasuredError(differences[n], predictedHarmonics[n], predictedAmplitudes[n], ampMax);
		}

        // measured to predicted
		differences = new Float32Array(peaks.length);
		for (let k = 0; k < differences.length; k++) {
			differences[k] = Number.POSITIVE_INFINITY;
			for (let n = 0; n < predictedHarmonics.length; n++) {
				let diff = Math.abs(predictedHarmonics[n] - peaks[k][0]);
				if(diff < differences[k]) {
					differences[k] = diff;
				}
			}
			measuredToPredictedError += this.computeMeasuredToPredictedError(differences[k], peaks[k][0], peaks[k][1], ampMax);
		}

		totalError = predictedToMeasuredError / predictedHarmonics.length + this.rho * measuredToPredictedError / peaks.length;
        return totalError;
    }

	computePredictedToMeasuredError(deltaFreqN, freqN, ampN, ampMax) {
		var error = deltaFreqN * Math.pow(freqN, -this.pmP) + (ampN / ampMax) * (this.pmQ * deltaFreqN * Math.pow(freqN, -this.pmP) - this.pmR);
		// freqN === 0 => error === Infinity
		return error;
	}

	computeMeasuredToPredictedError(deltaFreqK, freqK, ampK, ampMax) {
		var error = deltaFreqK * Math.pow(freqK, -this.mpP)+(ampK / ampMax) * (this.mpQ * deltaFreqK * Math.pow(freqK, -this.mpP) - this.mpR);
		return error;
	}
}

module.exports['TwoWayMismatch'] = TwoWayMismatch;
